### PR TITLE
Refactor Python provider detection and health checks

### PR DESCRIPTION
**Pull request by dlukes**
This is a follow-up on #11781, where I noted that [`s:check_python` may be unnecessarily complicated](https://github.com/neovim/neovim/pull/11781#issuecomment-579313377) (see https://github.com/neovim/neovim/issues/11753#issuecomment-578715584 for details) and asked whether there'd be interest in simplifying it. @justinmk said [yes, "as long as it results in a net reduction of code, and not a lot of indirection"](https://github.com/neovim/neovim/pull/11781#issuecomment-581191973).

Looking at the diff stats, the elephant in the room is that I actually ended up adding more code than I removed, but this is primarily because I ended up implementing reusable version comparison, including unit tests (see below). Hopefully adding tested code is less of a problem :) And if maintainers decide it should be removed, the PR will turn into a net reduction.

I realize there are quite a lot of code changes here and I'm obviously very open to feedback and dialing down on some of these ideas. I just thought they would be easier to discuss if I actually made the changes, rather than suggest / theorize about them. Please don't take this as an outsider stomping all over the codebase and lacking respect; take this as a conversation starter :)

### Demo and motivation

The main goal was to remove existing gnarliness and branching in Python provider detection and health checks, which currently leads to both false positives and false negatives.

Say you have the following setup:

```console
$ python3 -m venv venv
$ source venv/bin/activate
(venv) $ pip install pynvim
(venv) $ nvim -u NORC +'let g:python_host_prog = "python3" | checkhealth provider'
```

Installing `pynvim` inside the virtualenv simulates suboptimal setup for Python 3; setting the Python 2 host prog to `python3` simulates having an interpreter which doesn't satisfy version requirements.

Currently, this is the health check output:

image

The Python 2 provider gets set to a Python 3 interpreter (which can presumably break things) and just a warning is issued; the virtualenv problem slips by undetected (even though there is currently an error message for this type of situation, it just doesn't get triggered correctly).

With this PR, the health check would look something like this:

image

Trying to set a Python 3 interpreter as the Python 2 provider becomes a hard error; the virtualenv problem is reported on.

### High-level overview

`provider/pythonx`:

- shouldn't have a hardcoded list of Python executable names as possible candidates, it's a maintenance burden because new versions have to be added as they're released, and until they are, the provider will fail to detect them, which is inconvenient for early adopters of new Python releases who are on stable Neovim
- it should instead consider all viable candidates found on the `$PATH`, i.e. *not* just the first one: if the first `python3` found on the `$PATH` can't `import neovim`, it makes sense to try the remaining ones as well
- it should also return the full path to the provider interpreter (if one is successfully found) as well as its version, since it already has to determine them anyway

Then, `s:check_python` in `health/provider`:

- can entirely avoid having to figure out the full path to the provider interpreter on its own, which is error-prone and previously involved logic which didn't quite match that in `provider/pythonx`
- it can also avoid having to figure out the Python version (which is not that error-prone, but why do it twice)
- it can instead just focus on presenting correct troubleshooting messages

Additionally:

- I don't think Neovim should make a special effort to troubleshoot `pyenv`. It's hard to keep diagnostics in sync with a third-party tool. All that matters is getting a working Python binary (of the correct version); if `pyenv` shims get in the way, it's *their* job to fail with informative output, which they generally do, and Neovim makes sure the user sees it, which should be enough.
- I noticed some broken ad-hoc version comparisons while refactoring (using either vanilla string comparison or float/int conversions), so I decided to add a slightly more robust and reusable version comparison function to `lua-stdlib` and use that instead.

  When I was part way through implementing it, I noticed that a `s:version_cmp` function was used for this purpose in other places. I still saw benefit in having a public, more feature complete function with unit tests for this purpose, so I went ahead and finished it, and replaced `s:version_cmp` with it. This is very much up for debate though, of course.

I'm happy to discuss all of this further, either in this main thread or in threads attached to particular pieces of code.

### CODE REVIEW

File: runtime/autoload/health/provider.vim
Comment (brcolow) [inline runtime/autoload/health/provider.vim:282]: typo: environemnt => environment

Comment (dlukes) [reply runtime/autoload/health/provider.vim:282]: Well-spotted, thanks!

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:23]: Replaced with more elaborate function in `lua-stdlib` which has unit tests.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:186]: Not necessary, this information is now returned by `provider#pythonx#Detect`.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:258]: Not necessary, the Python interpreter returned by the provider is already guaranteed to be executable.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:244]: I thought the original message was somewhat confusing/unclear when reading the code -- the function is called `s:disabled_via_loaded_var`, the previous branch reports "Disabled" and returns 1, this one *also* reports "Disabled" but doesn't.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:310]: This is the safer way to invoke `pip` to make sure the right one is used.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:430]: These are some of the happy-go-lucky version checks I mentioned; they are no longer needed, `pythonx/provider` makes sure it only returns an appropriate Python 2 or 3 version that satisfies minimum version requirements.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:445]: This can never happen: in this branch, we're using a Python interpreter which has already proven it can `import neovim`.

File: runtime/autoload/health/provider.vim
Comment (dlukes) [inline runtime/autoload/health/provider.vim:460]: `pyenv` has no useful information to contribute regarding Python provider detection (apart from the possible error output captured in `provider/pythonx`), and as I've argued previously, I don't think it's worth it to maintain troubleshooting code for a third-party tool for no particular reason.

File: runtime/autoload/provider/pythonx.vim
Comment (dlukes) [inline runtime/autoload/provider/pythonx.vim:8]: I think it might be a good idea to put these minimum versions in a more prominent place at the beginning of the file.

File: runtime/autoload/provider/pythonx.vim
Comment (dlukes) [inline runtime/autoload/provider/pythonx.vim:100]: It's better to return a list of errors, the health check formats it prettily. And the "provider/pythonx: Could not load Python etc." message feels a bit redundant in the context that it's shown in.

File: runtime/autoload/provider/pythonx.vim
Comment (dlukes) [inline runtime/autoload/provider/pythonx.vim:109]: I wouldn't special case problems with `pyenv` shims, the error output should be enough for the user to determine what's going on, especially since Neovim's addition to it is just a cryptic "/foo/bar/python does not exist".

File: runtime/autoload/provider/pythonx.vim
Comment (dlukes) [inline runtime/autoload/provider/pythonx.vim:141]: This branch sounds like the user has a broken Python executable, I would encourage them to start by investigating whether they can even run it on the command line, before reporting an issue to Neovim.

File: test/functional/helpers.lua
Comment (dlukes) [inline test/functional/helpers.lua:776]: Because `provider#pythonx#Detect` now returns the full path to the provider interpreter, **its version**, and only then any errors.

**Comment by tjdevries** [comment]
Thanks for the PR. I will investigate this (but would like review from other maintainers if they're around for it)

**Comment by tjdevries** [comment]
One thought I had was now that Python 2 is end of life, I wonder if we should do anything differently with Python 2.

File: src/nvim/lua/vim.lua
Comment (tjdevries) [inline src/nvim/lua/vim.lua:504]: I don't like that sometimes this function returns a nicely formatted, informative map of semver versions and other times just returns a list of numbers.

I think we should just remove "non-strict" mode from this. If you just want to compare dotted numbers, it's fine to write the 4 lines of code you have above splitting the numbers. (Do we use non-strict mode anywhere internally)?

File: src/nvim/lua/vim.lua
Comment (tjdevries) [inline src/nvim/lua/vim.lua:531]: The strict mode comments go here as well.

One suggestion might be that there should be a "minimum comparison" key or similar. For example, most of the time I probably just want to check major, minor patch. If they are the same, then I'll call that "equal" -- even if they aren't strictly equal. I'm not sure exactly the best way to handle that though.

**Comment by tjdevries** [comment]
I'm not worried about how much code you wrote. 250 lines of the code are the implementation and testing of version compare, which is somewhat orthogonal to the original problem. So overall it looks like you succeeded in removing code ;) I will look some more later.

File: src/nvim/lua/vim.lua
Comment (dlukes) [reply src/nvim/lua/vim.lua:504]: I initially didn't have a non-strict mode, because I thought I'd just be replacing ad hoc string or numeric conversion based comparisons of Python versions where I had a good idea of the version string format. But then I came across the script-local function `s:version_cmp` in `health/provider`, which is used by some of the other provider checks and which just splits the version strings on periods, leniently converts the elements to numbers and compares them.

I concluded `s:version_cmp` should also be replaced by the new version comparison function, but suddenly I wasn't so sure about the range of inputs these additional provider checks would feed it. `s:version_cmp` can basically soldier through any kind of input, even if you pass in `'foo'` and `'bar'`, it won't give up. By contrast, the strict version of the new `cmp_versions` function doesn't allow more than three dotted numbers, because that's what semver syntax stipulates.

So I realized a couple of things:

1. Swapping out `s:version_cmp` for a strict `cmp_versions` might break some of the health checks, should they happen to involve version strings which don't conform to semver syntax (e.g. because they have some trailing junk, or because they have more than three numbers).
2. Environment detection and troubleshooting is hard to test because it involves external moving parts, so maybe there's value in having a more resilient mode for that.
3. In strict semver syntax mode, there's just no way to parse version strings consisting of more than three dotted numbers. Still, being able to do this can occasionally be useful, even within semver itself -- the pre-release label is fairly freeform, it can consist of any number of dotted numbers, among other things, so when `cmp_versions` reaches the point where it needs to call itself recursively to compare pre-release labels, it does so in non-strict mode for precisely this reason.

So I implemented a non-strict mode, which is still stricter than `s:version_cmp` because it at least expects some numbers with periods between them, which seems like a reasonable baseline (so it would choke on `'foo'` and `'bar'` as invalid version strings). I made it the default, but it could conceivably be made opt-in instead if that would help you warm up to the notion :)

If none of this sounds convincing -- it's entirely possible I'm being too cautious -- then I guess let's remove it. I'll try and check the format of the version strings returned by those other provider programs which are currently compared via `s:version_cmp`, make any adjustments necessary so that the comparisons in the health checks don't error out even under strict `cmp_versions`, and we'll hope it doesn't break in the future or if people use weird pre-release versions etc.

File: src/nvim/lua/vim.lua
Comment (dlukes) [reply src/nvim/lua/vim.lua:531]: Re: strict mode, please see https://github.com/neovim/neovim/pull/12437#discussion_r437754800 and let's have that discussion there?

> a "minimum comparison" key or similar

That's a good suggestion, thanks! I'll try and think of a way of incorporating it :)

**Comment by dlukes** [comment]
> now that Python 2 is end of life, I wonder if we should do anything differently with Python 2

Good point, I'm open to suggestions :) Will Neovim be deprecating Python 2 as a provider? Then the health checks could issue a deprecation warning.

**Comment by tjdevries** [comment]
Let's leave python 2 deprecation for a different time, although it would probably simplify quite a few things and I'm not sure how many people are even using the Python 2 plugin anyway... Let me know when you figure out exactly what you're looking to do for the semver stuff and I'll check this out again.

**Comment by dlukes** [comment]
> Let me know when you figure out exactly what you're looking to do for the semver stuff

I'm kind of waiting for feedback there :) I laid out [a rationale for adding and defaulting to a non-strict version comparison mode](https://github.com/neovim/neovim/pull/12437#discussion_r437754800), but I'm happy to rip it out if it doesn't sound convincing, it's entirely up to you.

Once this bigger issue is resolved, I thought I'd take a look at that ["minimum comparison" feature](https://github.com/neovim/neovim/pull/12437#discussion_r436690694) you mentioned.

**Review by justinmk — APPROVED** [review]
Unfortunate that this got lost. Looks like a good direction. Would have avoided https://github.com/neovim/neovim/pull/15937 

I wonder if the functions should have a common prefix, thus the names would be:

```
vim.version_cmp
vim.version_parse
```

**Comment by dlukes** [comment]
Thanks for digging this up!

> I wonder if the functions should have a common prefix

Sounds good to me :) Shall I add that, and revive the PR, when I get a moment?

What about [non-strict mode](https://github.com/neovim/neovim/pull/12437#discussion_r437754800) and ["minimum comparison" key](https://github.com/neovim/neovim/pull/12437#discussion_r436690694)?

**Comment by justinmk** [comment]
> Sounds good to me :) Shall I add that, and try to revive the PR, when I get a moment?

Would be much appreciated !

 
> What about [non-strict mode](https://github.com/neovim/neovim/pull/12437#discussion_r437754800) and ["minimum comparison" key](https://github.com/neovim/neovim/pull/12437#discussion_r436690694)?

1. at a glance, the features as you have them in this PR look pretty self-contained and are covered by tests so I don't see the extra feature as net cost.
2. "minimum comparison" OTOH seems like YAGNI, or at least a "two way door", i.e. we don't lose anything by waiting until later.

**Comment by RossSmyth** [comment]
If you see the linked issue, if this PR solves that it would be great. But rebasing this PR to main is more effort than I would like to put in to solving this issue as the patch I have in my issue works for me fine.

**Comment by dlukes** [comment]
@RossSmyth Rebased, could you please test if it addresses your issue if you get a moment?

**Comment by dlukes** [comment]
@tjdevries I'm sure you're swamped, but do you think you could take a look at this some time in the relatively near future? Or failing that, assign a different reviewer, so that we can figure out if this is going in or getting closed? :)

About the `news.txt` check: not sure if the correct resolution is adding the breaking change to `news.txt`, or just removing the mention from the commit message. While technically breaking, it's a relatively minor function that users are unlikely to call directly.

Also: there were quite a few changes from October 2021 when I last worked on this (e.g. dropping support for Python 2), the rebase was slightly unwieldy, and since it's diagnostic code, it has bad test coverage. If this gets merged and people subsequently report issues related to it, don't hesitate to @ me!

**Comment by RossSmyth** [comment]
Yeah I will check it out when I get home. Thanks!

**Comment by dlukes** [comment]
(I don't think the remaining CI failures -- apart from `news.txt` -- are on me, but maybe I'm not reading the logs correctly.)

**Comment by RossSmyth** [comment]
I couldn't get neovim to build. It's not you're fault, I couldn't get a tagged version to build either. I never use cmake and it seems whenever I do it just doesn't work so I don't know.

**Comment by dlukes** [comment]
@RossSmyth That's OK, no worries :) As I said, this PR simply removes Pyenv troubleshooting, the line you had to change to stop giving you trouble is gone, so that particular issue should be resolved.

**Comment by dlukes** [comment]
So while this PR was stalled waiting for feedback, Nvim apparently got a full Lua module for dealing with version comparisons: 0e7196438d8f856eecd7c90e160b79cbc8fb08dc

I'm happy to update the PR to reflect this (remove my own version comparison functions and tests, resolve the conflicts, update my code to use the version comparison API that's already been merged), but please let me know if there's still interest first :) With kids and all, I have little spare time, so I'm trying to choose carefully on what to spend it. @tjdevries, @justinmk?

**Comment by dundargoc** [comment]
This PR is massive and looks brutal to review. Are there unrelated improvements that we can split into a separate PRs perhaps?

**Comment by dlukes** [comment]
As I said, all of the Lua additions concerning version comparison can be removed, as they're superseded by the recently merged https://github.com/neovim/neovim/commit/0e7196438d8f856eecd7c90e160b79cbc8fb08dc. The rest changes how Python provider detection and health checks work to make it more resilient and correct, not sure if it can be sliced up further as these two things are quite intertwined.

**Comment by justinmk** [comment]
> The rest changes how Python provider detection and health checks work to make it more resilient and correct,

Those changes are still very much appreciated. But will need to adapt them to the new Lua-based implementation: https://github.com/neovim/neovim/pull/26941

**Pull request closed without merging**
